### DISCUSSION TITLE
nameof operator length parameter

### DISCUSSION THREAD

**Discussion by Pentadome** [dotnet/csharplang#3455] [Language Ideas]
nameof operator improvements
 Proposed

Summary
The nameof operator gets an additional optional parameter, where the amount of property names, to be converted to a string property path, can be indicated.

Motivation
Often in Xamarin, you want to bind to a property of an property. This means that you want to use the nameof operator to specify the property path. Currently this is done this way:
``` C#
nameof(SomethingViewModel.SomethingA) + "." +
nameof(SomethingViewModel.SomethingA.SomethingB)
```
This is a lot of boilerplate code for something that should be more simple.
Also, you might want to refer to a property of a property in an exception message.

Detailed design
nameof gets an additional parameter of type int where the length of the property path to convert to string can be indicated.

``` C#
nameof(SomethingViewModel.SomethingA) // "SomethingA"
nameof(SomethingViewModel.SomethingA, 1) // "SomethingA"

nameof(SomethingViewModel.SomethingA) + "." +
nameof(SomethingViewModel.SomethingA.SomethingB) // "SomethingA.SomethingB"
nameof(SomethingViewModel.SomethingA.SomethingB, 2)  // "SomethingA.SomethingB"

nameof(SomethingViewModel.SomethingA) + "." +
nameof(SomethingViewModel.SomethingA.SomethingB) + "." +
nameof(SomethingViewModel.SomethingA.SomethingB.SomethingC) //  "SomethingA.SomethingB.SomethingC"
nameof(SomethingViewModel.SomethingA.SomethingB.SomethingC, 3)  // "SomethingA.SomethingB.SomethingC"
nameof(SomethingViewModel.SomethingA.SomethingB.SomethingC, 0) // Optional, 0 means entire path. "SomethingA.SomethingB.SomethingC"
```

Drawbacks
At first glance, it might be unclear how long the property path is going to be,

Alternatives
The only alternative is provided in the proposal.

Unresolved questions
What happens when someone types in a higher number than the amount of properties?
What happens when someone uses a negative value?

**Comment by jnm2** [comment]
Related:
https://github.com/dotnet/csharplang/issues/701 "Proposal: fullnameof operator"
https://github.com/dotnet/csharplang/issues/3100 "Proposal: new operator like nameof, but for whole expression "

**Comment by emphasis87** [comment]
I wonder if array slicing could be used insted of start + length syntax.
